### PR TITLE
CORS-2291: terraform: save logs to a file

### PR DESCRIPTION
**Pull request by r4f4**
Just testing this within our CI structure.
/hold

### CODE REVIEW

**Comment by r4f4** [comment]
/retitle CORS-2291: terraform: save logs to a file

**Comment by r4f4** [comment]
On hold because of commit https://github.com/openshift/installer/pull/6532/commits/e9daa5b4a2ddc2bd455e97b2e1b23e511fb488db so we can see this PR in action in the CI runs.

**Comment by patrickdillon** [comment]
Nice work @r4f4 and @rna-afk! 

I didn't realize these terraform log files would be so massive... easily 50mb a pop.

**Comment by r4f4** [comment]
@patrickdillon I've added a commit to pipe through additional env vars, namely `TF_LOG_CORE` and `TF_LOG_PROVIDER` so we can select different debugging levels for each. Then we can play around with those in https://github.com/openshift/release/pull/32411 to find a middle ground where the logs are informative but not too gigantic.

**Comment by patrickdillon** [comment]
> @patrickdillon I've added a commit to pipe through additional env vars, namely `TF_LOG_CORE` and `TF_LOG_PROVIDER` so we can select different debugging levels for each. Then we can play around with those in [openshift/release#32411](https://github.com/openshift/release/pull/32411) to find a middle ground where the logs are informative but not too gigantic.

Ok. I have been trying, but with little luck to see if there are any unnecessary steps we can skip. The aws logs seem to be filled with useless lines repeating lines like this:
```
2022-10-28T15:59:36.782Z [DEBUG] provider.terraform-provider-aws: Called provider defined Type Validate: tf_req_id=91753a9d-d839-0ae7-408b-dbbd3ae6771f tf_rpc=PrepareProviderConfig @module=sdk.framework tf_attribute_path=endpoints[Value({"accessanalyzer":<null>,"account":<null>,"acm":<null>,"acmpca":<null>,"alexaforbusiness":<null>,"amg":<null>,"amp":<null>,"amplify":<null>,"amplifybackend":<null>,"amplifyuibuilder":<null>,"apigateway":<null>,"apigatewaymanagementapi":<null>,"apigatewayv2":<null>,"appautoscaling":<null>,"appconfig":<null>,"appconfigdata":<null>,"appflow":<null>,"appintegrations":<null>,"appintegrationsservice":<null>,"applicationautoscaling":<null>,"applicationcostprofiler":<null>,"applicationdiscovery":<null>,"applicationdiscoveryservice":<null>,"applicationinsights":<null>,"appmesh":<null>,"appregistry":<null>,"apprunner":<null>,"appstream":<null>,"appsync":<null>,"athena":<null>,"auditmanager":<null>,"augmentedairuntime":<null>,"autoscaling":<null>,"autoscalingplans":<null>,"backup":<null>,"backupgateway":<null>,"batch":<null>,"beanstalk":<null>,"billingconductor":<null>,"braket":<null>,"budgets":<null>,"ce":<null>,"chime":<null>,"chimesdkidentity":<null>,"chimesdkmeetings":<null>,"chimesdkmessaging":<null>,"cloud9":<null>,"cloudcontrol":<null>,"cloudcontrolapi":<null>,"clouddirectory":<null>,"cloudformation":<null>,"cloudfront":<null>,"cloudhsm":<null>,"cloudhsmv2":<null>,"cloudsearch":<null>,"cloudsearchdomain":<null>,"cloudtrail":<null>,"cloudwatch":<null>,"cloudwatchevents":<null>,"cloudwatchevidently":<null>,"cloudwatchlog":<null>,"cloudwatchlogs":<null>,"cloudwatchrum":<null>,"codeartifact":<null>,"codebuild":<null>,"codecommit":<null>,"codedeploy":<null>,"codeguruprofiler":<null>,"codegurureviewer":<null>,"codepipeline":<null>,"codestar":<null>,"codestarconnections":<null>,"codestarnotifications":<null>,"cognitoidentity":<null>,"cognitoidentityprovider":<null>,"cognitoidp":<null>,"cognitosync":<null>,"comprehend":<null>,"comprehendmedical":<null>,"computeoptimizer":<null>,"config":<null>,"configservice":<null>,"connect":<null>,"connectcontactlens":<null>,"connectparticipant":<null>,"connectwisdomservice":<null>,"controltower":<null>,"costandusagereportservice":<null>,"costexplorer":<null>,"cur":<null>,"customerprofiles":<null>,"databasemigration":<null>,"databasemigrationservice":<null>,"databrew":<null>,"dataexchange":<null>,"datapipeline":<null>,"datasync":<null>,"dax":<null>,"deploy":<null>,"detective":<null>,"devicefarm":<null>,"devopsguru":<null>,"directconnect":<null>,"directoryservice":<null>,"discovery":<null>,"dlm":<null>,"dms":<null>,"docdb":<null>,"drs":<null>,"ds":<null>,"dynamodb":<null>,"dynamodbstreams":<null>,"ebs":<null>,"ec2":<null>,"ec2instanceconnect":<null>,"ecr":<null>,"ecrpublic":<null>,"ecs":<null>,"efs":<null>,"eks":<null>,"elasticache":<null>,"elasticbeanstalk":<null>,"elasticinference":<null>,"elasticloadbalancing":<null>,"elasticloadbalancingv2":<null>,"elasticsearch":<null>,"elasticsearchservice":<null>,"elastictranscoder":<null>,"elb":<null>,"elbv2":<null>,"emr":<null>,"emrcontainers":<null>,"emrserverless":<null>,"es":<null>,"eventbridge":<null>,"events":<null>,"evidently":<null>,"finspace":<null>,"finspacedata":<null>,"firehose":<null>,"fis":<null>,"fms":<null>,"forecast":<null>,"forecastquery":<null>,"forecastqueryservice":<null>,"forecastservice":<null>,"frauddetector":<null>,"fsx":<null>,"gamelift":<null>,"glacier":<null>,"globalaccelerator":<null>,"glue":<null>,"gluedatabrew":<null>,"grafana":<null>,"greengrass":<null>,"greengrassv2":<null>,"groundstation":<null>,"guardduty":<null>,"health":<null>,"healthlake":<null>,"honeycode":<null>,"iam":<null>,"identitystore":<null>,"imagebuilder":<null>,"inspector":<null>,"inspector2":<null>,"inspectorv2":<null>,"iot":<null>,"iot1clickdevices":<null>,"iot1clickdevicesservice":<null>,"iot1clickprojects":<null>,"iotanalytics":<null>,"iotdata":<null>,"iotdataplane":<null>,"iotdeviceadvisor":<null>,"iotevents":<null>,"ioteventsdata":<null>,"iotfleethub":<null>,"iotjobsdata":<null>,"iotjobsdataplane":<null>,"iotsecuretunneling":<null>,"iotsitewise":<null>,"iotthingsgraph":<null>,"iottwinmaker":<null>,"iotwireless":<null>,"ivs":<null>,"kafka":<null>,"kafkaconnect":<null>,"kendra":<null>,"keyspaces":<null>,"kinesis":<null>,"kinesisanalytics":<null>,"kinesisanalyticsv2":<null>,"kinesisvideo":<null>,"kinesisvideoarchivedmedia":<null>,"kinesisvideomedia":<null>,"kinesisvideosignaling":<null>,"kinesisvideosignalingchannels":<null>,"kms":<null>,"lakeformation":<null>,"lambda":<null>,"lex":<null>,"lexmodelbuilding":<null>,"lexmodelbuildingservice":<null>,"lexmodels":<null>,"lexmodelsv2":<null>,"lexruntime":<null>,"lexruntimeservice":<null>,"lexruntimev2":<null>,"lexv2models":<null>,"lexv2runtime":<null>,"licensemanager":<null>,"lightsail":<null>,"location":<null>,"locationservice":<null>,"logs":<null>,"lookoutequipment":<null>,"lookoutforvision":<null>,"lookoutmetrics":<null>,"lookoutvision":<null>,"machinelearning":<null>,"macie":<null>,"macie2":<null>,"managedblockchain":<null>,"managedgrafana":<null>,"marketplacecatalog":<null>,"marketplacecommerceanalytics":<null>,"marketplaceentitlement":<null>,"marketplaceentitlementservice":<null>,"marketplacemetering":<null>,"mediaconnect":<null>,"mediaconvert":<null>,"medialive":<null>,"mediapackage":<null>,"mediapackagevod":<null>,"mediastore":<null>,"mediastoredata":<null>,"mediatailor":<null>,"memorydb":<null>,"meteringmarketplace":<null>,"mgh":<null>,"mgn":<null>,"migrationhub":<null>,"migrationhubconfig":<null>,"migrationhubrefactorspaces":<null>,"migrationhubstrategy":<null>,"migrationhubstrategyrecommendations":<null>,"mobile":<null>,"mq":<null>,"msk":<null>,"mturk":<null>,"mwaa":<null>,"neptune":<null>,"networkfirewall":<null>,"networkmanager":<null>,"nimble":<null>,"nimblestudio":<null>,"opensearch":<null>,"opensearchservice":<null>,"opsworks":<null>,"opsworkscm":<null>,"organizations":<null>,"outposts":<null>,"panorama":<null>,"personalize":<null>,"personalizeevents":<null>,"personalizeruntime":<null>,"pi":<null>,"pinpoint":<null>,"pinpointemail":<null>,"pinpointsmsvoice":<null>,"polly":<null>,"pricing":<null>,"prometheus":<null>,"prometheusservice":<null>,"proton":<null>,"qldb":<null>,"qldbsession":<null>,"quicksight":<null>,"ram":<null>,"rbin":<null>,"rds":<null>,"rdsdata":<null>,"rdsdataservice":<null>,"recyclebin":<null>,"redshift":<null>,"redshiftdata":<null>,"redshiftdataapiservice":<null>,"redshiftserverless":<null>,"rekognition":<null>,"resiliencehub":<null>,"resourcegroups":<null>,"resourcegroupstagging":<null>,"resourcegroupstaggingapi":<null>,"robomaker":<null>,"rolesanywhere":<null>,"route53":<null>,"route53domains":<null>,"route53recoverycluster":<null>,"route53recoverycontrolconfig":<null>,"route53recoveryreadiness":<null>,"route53resolver":<null>,"rum":<null>,"s3":<null>,"s3api":<null>,"s3control":<null>,"s3outposts":<null>,"sagemaker":<null>,"sagemakera2iruntime":<null>,"sagemakeredge":<null>,"sagemakeredgemanager":<null>,"sagemakerfeaturestoreruntime":<null>,"sagemakerruntime":<null>,"savingsplans":<null>,"schemas":<null>,"sdb":<null>,"secretsmanager":<null>,"securityhub":<null>,"serverlessapplicationrepository":<null>,"serverlessapprepo":<null>,"serverlessrepo":<null>,"servicecatalog":<null>,"servicecatalogappregistry":<null>,"servicediscovery":<null>,"servicequotas":<null>,"ses":<null>,"sesv2":<null>,"sfn":<null>,"shield":<null>,"signer":<null>,"simpledb":<null>,"sms":<null>,"snowball":<null>,"snowdevicemanagement":<null>,"sns":<null>,"sqs":<null>,"ssm":<null>,"ssmcontacts":<null>,"ssmincidents":<null>,"sso":<null>,"ssoadmin":<null>,"ssooidc":<null>,"stepfunctions":<null>,"storagegateway":<null>,"sts":<null>,"support":<null>,"swf":<null>,"synthetics":<null>,"textract":<null>,"timestreamquery":<null>,"timestreamwrite":<null>,"transcribe":<null>,"transcribeservice":<null>,"transcribestreaming":<null>,"transcribestreamingservice":<null>,"transfer":<null>,"translate":<null>,"voiceid":<null>,"waf":<null>,"wafregional":<null>,"wafv2":<null>,"wellarchitected":<null>,"wisdom":<null>,"workdocs":<null>,"worklink":<null>,"workmail":<null>,"workmailmessageflow":<null>,"workspaces":<null>,"workspacesweb":<null>,"xray":<null>})].cloudsearchdomain tf_mux_provider=*proto5server.Server tf_provider_addr=registry.terraform.io/hashicorp/aws @caller=/go/src/github.com/openshift/installer/terraform/providers/aws/vendor/github.com/hashicorp/terraform-plugin-framework/internal/fwschemadata/data_value.go:80 timestamp=2022-10-28T15:59:36.782Z
```

**Comment by r4f4** [comment]
I ran 2 deploys, one with `CORE=debug` and `PROVIDER=info` and vice-versa. The difference in log size is dramatic but I'm afraid we also lose all the useful info:

| CORE | PROVIDER | SIZE  | TAR.GZ |
| --- | --- | --- | --- |
| debug |  info             | 705K | 49K |
| info     | debug          | 62M  | 1.2M |
| debug | debug          | 63M  | 1.3M |

However, all repeating lines lend themselves well for compression, as can be seen from the `tar.gz` sizes above.
I have only looked into aws clusters. I'm assuming the provider log sizes will vary greatly from provider to provider.

**Comment by rna-afk** [comment]
I don't think we can get rid of useless lines without doing some post processing on the log files.

File: pkg/terraform/terraform.go [hunk 1]
```diff
@@ -25,14 +25,31 @@ func newTFExec(datadir string, terraformDir string) (*tfexec.Terraform, error) {
...
+	} else {
+		// Add terraform info logs to the installer log
+		lpDebug := &lineprinter.LinePrinter{Print: (&lineprinter.Trimmer{WrappedPrint: logrus.Debug}).Print}
+		lpError := &lineprinter.LinePrinter{Print: (&lineprinter.Trimmer{WrappedPrint: logrus.Error}).Print}
+		defer lpDebug.Close()
+		defer lpError.Close()
+
+		tf.SetStdout(lpDebug)
+		tf.SetStderr(lpError)
+		tf.SetLogger(newPrintfer())
```
Comment (r4f4) [inline pkg/terraform/terraform.go:51]: This is the part I'm not happy about. It seems the output we get from these is not the same we get from the log that goes into the file. I'm wondering if it's really a good idea to leave these out when we're sending tf logs to a file. Maybe we should still keep them in the `openshift_install.log`?

Comment (rna-afk) [reply pkg/terraform/terraform.go:51]: best to leave it to maximize the logs.

Comment (r4f4) [reply pkg/terraform/terraform.go]: Done

**Pull request merged**